In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
import warnings
warnings.filterwarnings('ignore')

In [2]:
df = pd.read_csv('cleaned_Concrete.csv')
df

cement   slag  flyash  superplasticizer  coarseaggregate  fineaggregate  \
0      540.0    0.0     0.0               2.5           1040.0          676.0   
1      540.0    0.0     0.0               2.5           1055.0          676.0   
2      332.5  142.5     0.0               0.0            932.0          594.0   
3      332.5  142.5     0.0               0.0            932.0          594.0   
4      198.6  132.4     0.0               0.0            978.4          825.5   
...      ...    ...     ...               ...              ...            ...   
1025   276.4  116.0    90.3               8.9            870.1          768.3   
1026   322.2    0.0   115.6              10.4            817.9          813.4   
1027   148.5  139.4   108.6               6.1            892.4          780.0   
1028   159.1  186.7     0.0              11.3            989.6          788.9   
1029   260.9  100.5    78.3               8.6            864.5          761.5   

      csMPa  boxcox_age  
0     79.99    3.619520  
1     61.89    3.619520  
2     40.27    6.441153  
3     41.05    6.840674  
4     44.30    6.822265  
...     ...         ...  
1025  44.28    3.619520  
1026  31.18    3.619520  
1027  23.70    3.619520  
1028  32.77    3.619520  
1029  32.40    3.619520  

[1030 rows x 8 columns]

In [3]:
X = df.drop(columns = {'csMPa'})
y = df['csMPa']

**Train Test Split**

In [4]:
from sklearn.model_selection import train_test_split

X_train, X_test, y_train, y_test = train_test_split(X, y, test_size = 0.2, random_state = 53)

In [5]:
#Creating Polynomials
from sklearn.preprocessing import PolynomialFeatures

poly = PolynomialFeatures(degree = 2, include_bias = False)
X_poly_train = pd.DataFrame(poly.fit_transform(X_train), columns = poly.get_feature_names_out())

#Modelling
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_poly_train, y_train)

#Prediction
ypred_train = model.predict(X_poly_train)

#Evaluation
from sklearn.metrics import r2_score

r2_train = r2_score(y_train, ypred_train)

from sklearn.model_selection import cross_val_score

cv = cross_val_score(model, X_poly_train, y_train, cv = 5, scoring = 'r2').mean()

print('Degree: 2')
print('R2(train): ', r2_train)
print('CV-Score: ', cv)

Degree: 2
R2(train):  0.8766404596361661
CV-Score:  0.8541771933671489


**Variable Significant**

In [6]:
X_poly_train

cement   slag  flyash  superplasticizer  coarseaggregate  fineaggregate  \
0     275.0    0.0     0.0               0.0           1088.0          808.0   
1     150.7    0.0   185.3              15.6           1074.5          678.0   
2     288.0  192.0     0.0               0.0            932.0          717.8   
3     190.3    0.0   125.2               9.9           1088.1          802.6   
4     255.0    0.0     0.0               0.0            889.8          945.0   
..      ...    ...     ...               ...              ...            ...   
819   491.0   26.0   123.0               3.9            822.0          699.0   
820   157.0  236.0     0.0               0.0            935.4          781.2   
821   178.0  129.8   118.6               3.6           1007.3          746.8   
822   153.0  145.0   113.0               8.0            867.0          824.0   
823   500.0    0.0     0.0               0.0           1125.0          613.0   

     boxcox_age   cement^2  cement slag  cement flyash  ...  \
0      3.619520   75625.00          0.0           0.00  ...   
1      3.619520   22710.49          0.0       27924.71  ...   
2      3.619520   82944.00      55296.0           0.00  ...   
3      5.165842   36214.09          0.0       23825.56  ...   
4      2.041652   65025.00          0.0           0.00  ...   
..          ...        ...          ...            ...  ...   
819    3.619520  241081.00      12766.0       60393.00  ...   
820    3.619520   24649.00      37052.0           0.00  ...   
821    2.817196   31684.00      23104.4       21110.80  ...   
822    3.619520   23409.00      22185.0       17289.00  ...   
823    5.913042  250000.00          0.0           0.00  ...   

     superplasticizer^2  superplasticizer coarseaggregate  \
0                  0.00                              0.00   
1                243.36                          16762.20   
2                  0.00                              0.00   
3                 98.01                          10772.19   
4                  0.00                              0.00   
..                  ...                               ...   
819               15.21                           3205.80   
820                0.00                              0.00   
821               12.96                           3626.28   
822               64.00                           6936.00   
823                0.00                              0.00   

     superplasticizer fineaggregate  superplasticizer boxcox_age  \
0                              0.00                     0.000000   
1                          10576.80                    56.464519   
2                              0.00                     0.000000   
3                           7945.74                    51.141841   
4                              0.00                     0.000000   
..                              ...                          ...   
819                         2726.10                    14.116130   
820                            0.00                     0.000000   
821                         2688.48                    10.141906   
822                         6592.00                    28.956163   
823                            0.00                     0.000000   

     coarseaggregate^2  coarseaggregate fineaggregate  \
0           1183744.00                      879104.00   
1           1154550.25                      728511.00   
2            868624.00                      668989.60   
3           1183961.61                      873309.06   
4            791744.04                      840861.00   
..                 ...                            ...   
819          675684.00                      574578.00   
820          874973.16                      730734.48   
821         1014653.29                      752251.64   
822          751689.00                      714408.00   
823         1265625.00                      689625.00   

     coarseaggregate boxcox_age  fineaggregate^2  fi

In [7]:
y_train = y_train.reset_index().drop(columns = {'index'})
y_train

csMPa
0    24.50
1    13.46
2    38.80
3    40.39
4    10.22
..     ...
819  57.92
820  33.66
821  34.24
822  26.23
823  51.04

[824 rows x 1 columns]

In [8]:
pd.set_option('display.max_columns', None)
df_poly_train = pd.concat([X_poly_train, y_train], axis = 1)
df_poly_train

cement   slag  flyash  superplasticizer  coarseaggregate  fineaggregate  \
0     275.0    0.0     0.0               0.0           1088.0          808.0   
1     150.7    0.0   185.3              15.6           1074.5          678.0   
2     288.0  192.0     0.0               0.0            932.0          717.8   
3     190.3    0.0   125.2               9.9           1088.1          802.6   
4     255.0    0.0     0.0               0.0            889.8          945.0   
..      ...    ...     ...               ...              ...            ...   
819   491.0   26.0   123.0               3.9            822.0          699.0   
820   157.0  236.0     0.0               0.0            935.4          781.2   
821   178.0  129.8   118.6               3.6           1007.3          746.8   
822   153.0  145.0   113.0               8.0            867.0          824.0   
823   500.0    0.0     0.0               0.0           1125.0          613.0   

     boxcox_age   cement^2  cement slag  cement flyash  \
0      3.619520   75625.00          0.0           0.00   
1      3.619520   22710.49          0.0       27924.71   
2      3.619520   82944.00      55296.0           0.00   
3      5.165842   36214.09          0.0       23825.56   
4      2.041652   65025.00          0.0           0.00   
..          ...        ...          ...            ...   
819    3.619520  241081.00      12766.0       60393.00   
820    3.619520   24649.00      37052.0           0.00   
821    2.817196   31684.00      23104.4       21110.80   
822    3.619520   23409.00      22185.0       17289.00   
823    5.913042  250000.00          0.0           0.00   

     cement superplasticizer  cement coarseaggregate  cement fineaggregate  \
0                       0.00               299200.00             222200.00   
1                    2350.92               161927.15             102174.60   
2                       0.00               268416.00             206726.40   
3                    1883.97               207065.43             152734.78   
4                       0.00               226899.00             240975.00   
..                       ...                     ...                   ...   
819                  1914.90               403602.00             343209.00   
820                     0.00               146857.80             122648.40   
821                   640.80               179299.40             132930.40   
822                  1224.00               132651.00             126072.00   
823                     0.00               562500.00             306500.00   

     cement boxcox_age    slag^2  slag flyash  slag superplasticizer  \
0           995.368117      0.00         0.00                   0.00   
1           545.461728      0.00         0.00                   0.00   
2          1042.421882  36864.00         0.00                   0.00   
3           983.059823      0.00         0.00                   0.00   
4           520.621140      0.00         0.00                   0.00   
..                 ...       ...          ...                    ...   
819        1777.184528    676.00      3198.00                 101.40   
820         568.264707  55696.00         0.00                   0.00   
821         501.460929  16848.04     15394.28                 467.28   
822         553.786625  21025.00     16385.00                1160.00   
823        2956.520928      0.00         0.00                   0.00   

     slag coarseaggregate  slag fineaggregate  slag boxcox_age  flyash^2  \
0                    0.00                0.00         0.000000      0.00   
1                    0.00                0.00         0.000000  34336.09   
2               178944.00           137817.60       694.947922      0.00   
3                    0.00                0.00         0.000000  15675.04   
4                    0.00                0.00         0.000000      0.00   
..                    ...                 ...              ...       ...   
819          

In [9]:
import statsmodels.formula.api as smf

model = smf.ols('y_train~X_poly_train', data = df_poly_train).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                y_train   R-squared:                       0.877
Model:                            OLS   Adj. R-squared:                  0.871
Method:                 Least Squares   F-statistic:                     160.0
Date:                Tue, 11 Feb 2025   Prob (F-statistic):               0.00
Time:                        13:10:50   Log-Likelihood:                -2621.6
No. Observations:                 824   AIC:                             5315.
Df Residuals:                     788   BIC:                             5485.
Df Model:                          35                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept         1116.1948    198.154      5.633      0.000     727.223    1505.167
X_poly_train[0]     -0.9899      0.205     -4.829      0.000      -1.392      -0.588
X_poly_train[1]     -1.0673      0.245     -4.361      0.000      -1.548      -0.587
X_poly_train[2]     -1.9848      0.334     -5.950      0.000      -2.640      -1.330
X_poly_train[3]     -3.1030      2.371     -1.309      0.191      -7.757       1.551
X_poly_train[4]     -1.2369      0.197     -6.294      0.000      -1.623      -0.851
X_poly_train[5]     -0.7825      0.194     -4.025      0.000      -1.164      -0.401
X_poly_train[6]    -17.4080      6.408     -2.717      0.007     -29.986      -4.830
X_poly_train[7]      0.0002   6.12e-05      3.200      0.001    7.58e-05       0.000
X_poly_train[8]      0.0005      0.000      3.115      0.002       0.000       0.001
X_poly_train[9]      0.0009      0.000      4.459      0.000       0.000       0.001
X_poly_train[10]     0.0013      0.001      1.023      0.306      -0.001       0.004
X_poly_train[11]     0.0006   9.05e-05      6.478      0.000       0.000       0.001
X_poly_train[12]     0.0004      0.000      4.150      0.000       0.000       0.001
X_poly_train[13]     0.0150      0.003      4.282      0.000       0.008       0.022
X_poly_train[14]     0.0002   9.27e-05      1.882      0.060   -7.48e-06       0.000
X_poly_train[15]     0.0009      0.000      3.825      0.000       0.000       0.001
X_poly_train[16]     0.0016      0.002      0.973      0.331      -0.002       0.005
X_poly_train[17]     0.0005      0.000      4.803      0.000       0.000       0.001
X_poly_train[18]     0.0005      0.000      3.912      0.000       0.000       0.001
X_poly_train[19]     0.0278      0.004      6.757      0.000       0.020       0.036
X_poly_train[20]     0.0006      0.000      3.286      0.001       0.000       0.001
X_poly_train[21]    -0.0009      0.002     -0.388      0.698      -0.005       0.004
X_poly_train[22]     0.0009      0.000      6.209      0.000       0.001       0.001
X_poly_train[23]     0.0009      0.000      5.374      0.000       0.001       0.001
X_poly_train[24]     0.0280      0.005      5.266      0.000       0.018       0.038
X_poly_train[25]    -0.0255      0.008     -3.178      0.002      -0.041      -0.010
X_poly_train[26]     0.0029      0.001      2.537      0.011       0.001       0.005
X_poly_train[27]     0.0010      0.001      0.796      0.426      -0.001       0.003
X_poly_train[28]    -0.0347      0.043     -0.801      0.423      -0.120       0.050
X_poly_train[29]     0.0003    5.6e-05      5.143      0.000       0.000       0.000
X_poly_train[30]     0.0005   8.79e-05      6.148      0.000       0.000       0.001
X_poly_train[31]     0.0116      0.003      3.807      0.000       0.006       0.018
X_poly_train[32]  4.101

-------------------------------------------------------------------------------

**So, index {3, 10, 14, 16, 21, 27, 28, 32} --> p > 0.05**

Index 21 column (flyash superplasticizer) has greatest p value --> p = 0.698 --> Let's Drop it first

In [10]:
# df_poly_train = df_poly_train.drop(columns = {'flyash superplasticizer'})
df_poly_train = df_poly_train.drop(df_poly_train.columns[21], axis=1)

In [11]:
X_poly_train = df_poly_train.drop(columns = {'csMPa'})
y = df_poly_train['csMPa']

In [12]:
#Modelling
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_poly_train, y_train)

#Prediction
ypred_train = model.predict(X_poly_train)

#Evaluation
from sklearn.metrics import r2_score

r2_train = r2_score(y_train, ypred_train)

from sklearn.model_selection import cross_val_score

cv = cross_val_score(model, X_poly_train, y_train, cv = 5, scoring = 'r2').mean()

print('Degree: 2')
print('R2(train): ', r2_train)
print('CV-Score: ', cv)

Degree: 2
R2(train):  0.8766169435171227
CV-Score:  0.8554855912736127


In [13]:
import statsmodels.formula.api as smf

model = smf.ols('y_train~X_poly_train', data = df_poly_train).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                y_train   R-squared:                       0.877
Model:                            OLS   Adj. R-squared:                  0.871
Method:                 Least Squares   F-statistic:                     164.9
Date:                Tue, 11 Feb 2025   Prob (F-statistic):               0.00
Time:                        13:10:50   Log-Likelihood:                -2621.7
No. Observations:                 824   AIC:                             5313.
Df Residuals:                     789   BIC:                             5478.
Df Model:                          34                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept         1092.6529    188.513      5.796      0.000     722.606    1462.700
X_poly_train[0]     -0.9588      0.189     -5.086      0.000      -1.329      -0.589
X_poly_train[1]     -1.0289      0.224     -4.599      0.000      -1.468      -0.590
X_poly_train[2]     -1.9128      0.277     -6.909      0.000      -2.456      -1.369
X_poly_train[3]     -3.8243      1.468     -2.605      0.009      -6.707      -0.942
X_poly_train[4]     -1.2230      0.193     -6.332      0.000      -1.602      -0.844
X_poly_train[5]     -0.7553      0.181     -4.169      0.000      -1.111      -0.400
X_poly_train[6]    -17.1552      6.371     -2.693      0.007     -29.661      -4.649
X_poly_train[7]      0.0002   5.82e-05      3.242      0.001    7.44e-05       0.000
X_poly_train[8]      0.0004      0.000      3.325      0.001       0.000       0.001
X_poly_train[9]      0.0008      0.000      5.419      0.000       0.001       0.001
X_poly_train[10]     0.0018      0.001      3.051      0.002       0.001       0.003
X_poly_train[11]     0.0006   8.53e-05      6.737      0.000       0.000       0.001
X_poly_train[12]     0.0004   9.51e-05      4.391      0.000       0.000       0.001
X_poly_train[13]     0.0148      0.003      4.267      0.000       0.008       0.022
X_poly_train[14]     0.0002   8.46e-05      1.889      0.059   -6.26e-06       0.000
X_poly_train[15]     0.0009      0.000      4.556      0.000       0.001       0.001
X_poly_train[16]     0.0022      0.001      2.741      0.006       0.001       0.004
X_poly_train[17]     0.0005      0.000      4.864      0.000       0.000       0.001
X_poly_train[18]     0.0005      0.000      4.190      0.000       0.000       0.001
X_poly_train[19]     0.0277      0.004      6.759      0.000       0.020       0.036
X_poly_train[20]     0.0006      0.000      4.143      0.000       0.000       0.001
X_poly_train[21]     0.0009      0.000      6.676      0.000       0.001       0.001
X_poly_train[22]     0.0009      0.000      6.129      0.000       0.001       0.001
X_poly_train[23]     0.0277      0.005      5.262      0.000       0.017       0.038
X_poly_train[24]    -0.0271      0.007     -3.998      0.000      -0.040      -0.014
X_poly_train[25]     0.0031      0.001      3.340      0.001       0.001       0.005
X_poly_train[26]     0.0013      0.001      1.611      0.108      -0.000       0.003
X_poly_train[27]    -0.0335      0.043     -0.775      0.439      -0.118       0.051
X_poly_train[28]     0.0003   5.58e-05      5.131      0.000       0.000       0.000
X_poly_train[29]     0.0005   8.53e-05      6.241      0.000       0.000       0.001
X_poly_train[30]     0.0116      0.003      3.796      0.000       0.006       0.018
X_poly_train[31]  3.309e-05   5.48e-05      0.604      0.546   -7.45e-05       0.000
X_poly_train[32]     0.

**So, index {14, 26, 27, 31} --> p > 0.05**

Index 31 column (fineaggregate^2) has greatest p value --> p = 0.546 --> Let's Drop it first

In [14]:
df_poly_train = df_poly_train.drop(df_poly_train.columns[31], axis = 1)

In [15]:
X_poly_train = df_poly_train.drop(columns = {'csMPa'})
y = df_poly_train['csMPa']

In [16]:
#Modelling
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_poly_train, y_train)

#Prediction
ypred_train = model.predict(X_poly_train)

#Evaluation
from sklearn.metrics import r2_score

r2_train = r2_score(y_train, ypred_train)

from sklearn.model_selection import cross_val_score

cv = cross_val_score(model, X_poly_train, y_train, cv = 5, scoring = 'r2').mean()

print('Degree: 2')
print('R2(train): ', r2_train)
print('CV-Score: ', cv)

Degree: 2
R2(train):  0.8765599593236217
CV-Score:  0.8557391068346927


In [17]:
import statsmodels.formula.api as smf

model = smf.ols('y_train~X_poly_train', data = df_poly_train).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                y_train   R-squared:                       0.877
Model:                            OLS   Adj. R-squared:                  0.871
Method:                 Least Squares   F-statistic:                     170.0
Date:                Tue, 11 Feb 2025   Prob (F-statistic):               0.00
Time:                        13:10:51   Log-Likelihood:                -2621.9
No. Observations:                 824   AIC:                             5312.
Df Residuals:                     790   BIC:                             5472.
Df Model:                          33                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept         1020.3250    145.479      7.014      0.000     734.754    1305.896
X_poly_train[0]     -0.8963      0.157     -5.693      0.000      -1.205      -0.587
X_poly_train[1]     -0.9476      0.179     -5.306      0.000      -1.298      -0.597
X_poly_train[2]     -1.8269      0.237     -7.696      0.000      -2.293      -1.361
X_poly_train[3]     -4.2297      1.305     -3.240      0.001      -6.792      -1.667
X_poly_train[4]     -1.1802      0.180     -6.572      0.000      -1.533      -0.828
X_poly_train[5]     -0.6590      0.086     -7.657      0.000      -0.828      -0.490
X_poly_train[6]    -15.9979      6.073     -2.634      0.009     -27.919      -4.077
X_poly_train[7]      0.0002   5.38e-05      3.260      0.001    6.97e-05       0.000
X_poly_train[8]      0.0004      0.000      3.421      0.001       0.000       0.001
X_poly_train[9]      0.0008      0.000      5.645      0.000       0.001       0.001
X_poly_train[10]     0.0018      0.001      3.179      0.002       0.001       0.003
X_poly_train[11]     0.0006      8e-05      6.961      0.000       0.000       0.001
X_poly_train[12]     0.0004   6.56e-05      5.736      0.000       0.000       0.001
X_poly_train[13]     0.0144      0.003      4.235      0.000       0.008       0.021
X_poly_train[14]     0.0001   7.43e-05      1.822      0.069   -1.05e-05       0.000
X_poly_train[15]     0.0008      0.000      4.754      0.000       0.000       0.001
X_poly_train[16]     0.0023      0.001      3.055      0.002       0.001       0.004
X_poly_train[17]     0.0005   9.85e-05      4.994      0.000       0.000       0.001
X_poly_train[18]     0.0004   7.46e-05      5.821      0.000       0.000       0.001
X_poly_train[19]     0.0271      0.004      6.789      0.000       0.019       0.035
X_poly_train[20]     0.0006      0.000      4.104      0.000       0.000       0.001
X_poly_train[21]     0.0009      0.000      6.802      0.000       0.001       0.001
X_poly_train[22]     0.0008      0.000      7.761      0.000       0.001       0.001
X_poly_train[23]     0.0272      0.005      5.232      0.000       0.017       0.037
X_poly_train[24]    -0.0271      0.007     -3.999      0.000      -0.040      -0.014
X_poly_train[25]     0.0033      0.001      3.615      0.000       0.001       0.005
X_poly_train[26]     0.0016      0.001      2.467      0.014       0.000       0.003
X_poly_train[27]    -0.0302      0.043     -0.704      0.481      -0.114       0.054
X_poly_train[28]     0.0003   5.45e-05      5.122      0.000       0.000       0.000
X_poly_train[29]     0.0005   7.29e-05      6.938      0.000       0.000       0.001
X_poly_train[30]     0.0111      0.003      3.767      0.000       0.005       0.017
X_poly_train[31]     0.0123      0.003      3.586      0.000       0.006       0.019
X_poly_train[32]    -0.

**So, index {6, 14} --> p > 0.05**

Index 6 column has greatest p value --> p = 0.229 --> Let's Drop it first

In [18]:
df_poly_train = df_poly_train.drop(df_poly_train.columns[6], axis = 1)

In [19]:
X_poly_train = df_poly_train.drop(columns = {'csMPa'})
y_train = df_poly_train['csMPa']

In [20]:
#Modelling
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_poly_train, y_train)

#Prediction
ypred_train = model.predict(X_poly_train)

#Evaluation
from sklearn.metrics import r2_score

r2_train = r2_score(y_train, ypred_train)

from sklearn.model_selection import cross_val_score

cv = cross_val_score(model, X_poly_train, y_train, cv = 5, scoring = 'r2').mean()

print('Degree: 2')
print('R2(train): ', r2_train)
print('CV-Score: ', cv)

Degree: 2
R2(train):  0.8754757111365741
CV-Score:  0.8567343587673207


In [21]:
import statsmodels.formula.api as smf

model = smf.ols('y_train~X_poly_train', data = df_poly_train).fit()
model.summary()

<class 'statsmodels.iolib.summary.Summary'>
"""
                            OLS Regression Results                            
==============================================================================
Dep. Variable:                y_train   R-squared:                       0.875
Model:                            OLS   Adj. R-squared:                  0.870
Method:                 Least Squares   F-statistic:                     173.8
Date:                Tue, 11 Feb 2025   Prob (F-statistic):               0.00
Time:                        13:10:51   Log-Likelihood:                -2625.5
No. Observations:                 824   AIC:                             5317.
Df Residuals:                     791   BIC:                             5473.
Df Model:                          32                                         
Covariance Type:            nonrobust                                         
====================================================================================
                       coef    std err          t      P>|t|      [0.025      0.975]
------------------------------------------------------------------------------------
Intercept          920.4385    140.977      6.529      0.000     643.706    1197.171
X_poly_train[0]     -0.8292      0.156     -5.317      0.000      -1.135      -0.523
X_poly_train[1]     -0.8667      0.177     -4.908      0.000      -1.213      -0.520
X_poly_train[2]     -1.7207      0.235     -7.329      0.000      -2.182      -1.260
X_poly_train[3]     -4.2896      1.310     -3.275      0.001      -6.861      -1.718
X_poly_train[4]     -1.1164      0.179     -6.251      0.000      -1.467      -0.766
X_poly_train[5]     -0.5990      0.083     -7.190      0.000      -0.763      -0.435
X_poly_train[6]      0.0002   5.38e-05      3.075      0.002    5.99e-05       0.000
X_poly_train[7]      0.0004      0.000      3.187      0.001       0.000       0.001
X_poly_train[8]      0.0007      0.000      5.404      0.000       0.000       0.001
X_poly_train[9]      0.0019      0.001      3.243      0.001       0.001       0.003
X_poly_train[10]     0.0005      8e-05      6.756      0.000       0.000       0.001
X_poly_train[11]     0.0004   6.54e-05      5.442      0.000       0.000       0.000
X_poly_train[12]     0.0073      0.002      3.526      0.000       0.003       0.011
X_poly_train[13]     0.0001   7.44e-05      1.627      0.104    -2.5e-05       0.000
X_poly_train[14]     0.0008      0.000      4.504      0.000       0.000       0.001
X_poly_train[15]     0.0024      0.001      3.120      0.002       0.001       0.004
X_poly_train[16]     0.0005   9.86e-05      4.788      0.000       0.000       0.001
X_poly_train[17]     0.0004   7.43e-05      5.521      0.000       0.000       0.001
X_poly_train[18]     0.0186      0.002      7.904      0.000       0.014       0.023
X_poly_train[19]     0.0005      0.000      3.871      0.000       0.000       0.001
X_poly_train[20]     0.0009      0.000      6.595      0.000       0.001       0.001
X_poly_train[21]     0.0008      0.000      7.423      0.000       0.001       0.001
X_poly_train[22]     0.0191      0.004      4.540      0.000       0.011       0.027
X_poly_train[23]    -0.0268      0.007     -3.943      0.000      -0.040      -0.013
X_poly_train[24]     0.0032      0.001      3.524      0.000       0.001       0.005
X_poly_train[25]     0.0015      0.001      2.236      0.026       0.000       0.003
X_poly_train[26]     0.0313      0.036      0.867      0.386      -0.040       0.102
X_poly_train[27]     0.0003   5.46e-05      4.954      0.000       0.000       0.000
X_poly_train[28]     0.0005   7.28e-05      6.686      0.000       0.000       0.001
X_poly_train[29]     0.0047      0.002      2.806      0.005       0.001       0.008
X_poly_train[30]     0.0043      0.002      2.708      0.007       0.001       0.007
X_poly_train[31]    -0.7172      0.100     -7.207      0.000      -0.913      -0.522
=======================

**So, index {13} --> p > 0.05** --> Let's drop it

Dropping more columns are leading to increase in p for different columns --> so let's stop here and this is our final model for degree = 2

# **Final Model**

In [22]:
X_poly_test = pd.DataFrame(poly.fit_transform(X_test), columns = poly.get_feature_names_out())

In [23]:
y_test = y_test.reset_index().drop(columns = {'index'})

In [24]:
df_poly_test = pd.concat([X_poly_test, y_test], axis = 1)
df_poly_test

cement   slag  flyash  superplasticizer  coarseaggregate  fineaggregate  \
0     136.0  162.0   126.0              10.0            923.0          764.0   
1     304.0  140.0     0.0               6.0            895.0          722.0   
2     155.0  183.0     0.0               9.0           1047.0          697.0   
3     203.5  135.7     0.0               0.0           1076.2          759.3   
4     393.0    0.0     0.0               0.0            940.6          785.6   
..      ...    ...     ...               ...              ...            ...   
201   425.0  106.3     0.0              16.5            852.1          887.1   
202   331.0    0.0     0.0               0.0            978.0          825.0   
203   255.5  170.3     0.0               0.0           1026.6          724.3   
204   145.0    0.0   134.0              11.0            979.0          812.0   
205   339.0    0.0     0.0               0.0           1069.0          754.0   

     boxcox_age   cement^2  cement slag  cement flyash  \
0      3.619520   18496.00     22032.00        17136.0   
1      3.619520   92416.00     42560.00            0.0   
2      3.619520   24025.00     28365.00            0.0   
3      3.619520   41412.25     27614.95            0.0   
4      1.128705  154449.00         0.00            0.0   
..          ...        ...          ...            ...   
201    3.619520  180625.00     45177.50            0.0   
202    2.041652  109561.00         0.00            0.0   
203    3.619520   65280.25     43511.65            0.0   
204    3.619520   21025.00         0.00        19430.0   
205    2.041652  114921.00         0.00            0.0   

     cement superplasticizer  cement coarseaggregate  cement fineaggregate  \
0                     1360.0                125528.0             103904.00   
1                     1824.0                272080.0             219488.00   
2                     1395.0                162285.0             108035.00   
3                        0.0                219006.7             154517.55   
4                        0.0                369655.8             308740.80   
..                       ...                     ...                   ...   
201                   7012.5                362142.5             377017.50   
202                      0.0                323718.0             273075.00   
203                      0.0                262296.3             185058.65   
204                   1595.0                141955.0             117740.00   
205                      0.0                362391.0             255606.00   

     cement boxcox_age    slag^2  slag flyash  slag superplasticizer  \
0           492.254778  26244.00      20412.0                1620.00   
1          1100.334209  19600.00          0.0                 840.00   
2           561.025666  33489.00          0.0                1647.00   
3           736.572406  18414.49          0.0                   0.00   
4           443.581049      0.00          0.0                   0.00   
..                 ...       ...          ...                    ...   
201        1538.296180  11299.69          0.0                1753.95   
202         675.786656      0.00          0.0                   0.00   
203         924.787468  29002.09          0.0                   0.00   
204         524.830462      0.00          0.0                   0.00   
205         692.119868      0.00          0.0                   0.00   

     slag coarseaggregate  slag fineaggregate  slag boxcox_age  flyash^2  \
0               149526.00           123768.00       586.362309   15876.0   
1               125300.00           101080.00       506.732859       0.0   
2               191601.00           127551.00       662.372238       0.0   
3               146040.34           103037.01       491.168922       0.0   
4                    0.00                0.00         0.000000       0.0   
..                    ...                 ...              ...       ...   
201          

In [25]:
#Dropping all the unimportant columns here
df_poly_test = df_poly_test.drop(df_poly_test.columns[21], axis=1)
df_poly_test = df_poly_test.drop(df_poly_test.columns[31], axis = 1)
df_poly_test = df_poly_test.drop(df_poly_test.columns[6], axis = 1)

In [26]:
X_poly_test = df_poly_test.drop(columns = {'csMPa'})
y_test = df_poly_test['csMPa']

In [27]:
#Modelling
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X_poly_train, y_train)

#Prediction
ypred_train = model.predict(X_poly_train)
ypred_test = model.predict(X_poly_test)

#Evaluation
from sklearn.metrics import r2_score

r2_train = r2_score(y_train, ypred_train)
r2_test = r2_score(y_test, ypred_test)

from sklearn.model_selection import cross_val_score

cv = cross_val_score(model, X_poly_train, y_train, cv = 5, scoring = 'r2').mean()

print('Degree: 2')
print('R2(train): ', r2_train)
print('CV-Score: ', cv)
print('R2(test): ', r2_test)

if (abs(cv - r2_train) <= 0.05) and (abs(r2_train - r2_test) <= 0.05):
  print('Good Model')
else:
  print('Bad Model')

Degree: 2
R2(train):  0.8754757111365741
CV-Score:  0.8567343587673207
R2(test):  0.8658981823112474
Good Model
